src/unistd.rs

```diff
@@ -3151,6 +3151,139 @@ impl User {
             unsafe { libc::getpwnam_r(name.as_ptr(), pwd, cbuf, cap, res) }
         })
     }
+
+    /// Construct an iterator over all users on the system.
+    ///
+    /// Call [`setpwent(3)`](https://pubs.opengroup.org/onlinepubs/7908799/xsh/setpwent.html),
+    /// [`getpwent(3)`](https://pubs.opengroup.org/onlinepubs/7908799/xsh/getpwent.html),
+    /// [`getpwent_r(3)`](https://man7.org/linux/man-pages/man3/getpwent_r.3.html) and
+    /// [`endpwent(3)`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/endpwent.html)
+    /// under the hood.
+    ///
+    /// When encounters an error, [`AllUsers::next()`] will yield a
+    /// `Some(Err(Errno))`, stop iterating if this happens.
+    ///
+    /// # Safety
+    ///
+    /// This function is marked as `unsafe` cause `setpwent()` and `endpwent()`
+    /// will modify a global state, which will result in a data race if we have
+    /// multiple instances running at the same time.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use nix::unistd::User;
+    /// for opt_user in unsafe { User::all_users() } {
+    ///     match opt_user {
+    ///         Ok(user) => println!("{:?}", user),
+    ///         Err(errno) => {
+    ///             eprintln!("{:?}", errno);
+    ///             break;
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    // RedoxFS does not support passwd and android does not have /etc/passwd.
+    #[cfg(not(any(target_os = "redox", target_os = "android")))]
+    pub unsafe fn all_users() -> AllUsers {
```

Comment:
I suggest that a more natural API would be to call this function `AllUsers::new()`